glean-core/src/metrics/labeled.rs

```diff
@@ -0,0 +1,118 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashSet;
+use std::marker::PhantomData;
+
+use lazy_static::lazy_static;
+use regex::Regex;
+
+use crate::metrics::MetricType;
+use crate::CommonMetricData;
+use crate::Glean;
+
+const MAX_LABELS: usize = 16;
+const OTHER_LABEL: &str = "__other__";
+const MAX_LABEL_LENGTH: usize = 30;
+lazy_static! {
+    static ref LABEL_REGEX: Regex = Regex::new("^[a-z_][a-z0-9_]{0,29}$").unwrap();
+}
+
+/// A labeled metric.
+///
+/// Labeled metrics allow to record multiple sub-metrics of the same type under different string labels.
+#[derive(Debug)]
+pub struct LabeledMetric<T> {
+    meta: CommonMetricData,
+    labels: Option<Vec<String>>,
+    /// Type of the underlying metric
+    /// We don't hold on to any of these, but we need to be able to create them.
+    submetric: PhantomData<T>,
```

Comment:
This works because we return a new instance of the submetric, we only need to know its type.
On the Rust side we create a new submetric on `get`. 
On the Kotlin side this is kept around (but also destroyed in the finalizer).